libraries/AP_SmartAudio/AP_SmartAudio.cpp

```diff
@@ -0,0 +1,881 @@
+        set_channel((AP::vtx().get_band()*8)+AP::vtx().get_channel());
+    }
+
+    // send request update for power with ap_vtx values
+    if (current_state.version==SMARTAUDIO_SPEC_PROTOCOL_v21 && _get_power_in_mw_from_dbm(current_state.power_in_dbm)!= AP::vtx().get_power_mw()){
+        debug("UPDATE AP_VTX->HW_VTX: POW IN MW FROM DBM");
+        set_power_mw(AP::vtx().get_power_mw());
+    }
+
+    // send request update for power with ap_vtx values
+    if (current_state.version!=SMARTAUDIO_SPEC_PROTOCOL_v21
+    && AP::vtx().get_power_mw()!=_get_power_in_mw_from_dbm(_get_power_in_dbm_from_vtx_power_level(current_state.power, current_state.version))){
+        debug("UPDATE AP_VTX->HW_VTX: POW IN MW FROM PWLEVEL %d mw -> %d mw"
+        , AP::vtx().get_power_mw()
+        , _get_power_in_mw_from_dbm(_get_power_in_dbm_from_vtx_power_level(current_state.power, current_state.version))
+        );
+
+        set_power_mw(AP::vtx().get_power_mw());
+    }
+
+
+
+
+    uint8_t curr_operation_mode=(current_state.pitmodeInRangeActive<<0)
+    | (current_state.pitmodeOutRangeActive<<1)
+    | (current_state.pitModeRunning<<2)
+    | ((current_state.unlocked>0)<<3);
+
+    uint8_t new_operation_mode=curr_operation_mode;
+
+    new_operation_mode ^= (-(AP::vtx().get_options() & uint8_t(AP_VideoTX::VideoOptions::VTX_PITMODE)) ^ new_operation_mode) & (1UL << 2);
+
+    new_operation_mode ^= (-(AP::vtx().get_options() & uint8_t(AP_VideoTX::VideoOptions::VTX_UNLOCKED)) ^ new_operation_mode) & (1UL << 3);
+
+    if(curr_operation_mode!=new_operation_mode){
+        debug("UPDATE AP_VTX->HW_VTX: OPTIONS");
+        set_operation_mode(new_operation_mode);
+    }
+
+    return true;
+}
+
+/**
+ * Sends an SmartAudio Command to the vtx, waits response on the update event
+ * @param frameBuffer frameBuffer to send over the wire
+ * @param size  size of the framebuffer wich needs to be sended
+ */
+void AP_SmartAudio::send_request(smartaudioFrame_t requestFrame, uint8_t size)
+{
+
+    if (size<=0) {
+        debug("ERROR - %s HW: %s", TAG, "CANNOT SEND REQUEST, REQUEST IS EMPTY");
+        return;
+    }
+    if (_port==nullptr) {
+        debug("ERROR - %s HW: %s", TAG, "CANNOT SEND REQUEST, UART NOT CONNECTED");
+        return;
+    }
+
+    uint8_t *request=reinterpret_cast<uint8_t*>(&requestFrame.u8RequestFrame);
+    // pull line low
+    _port->write((uint8_t)0);
+
+    // write request
+    for (uint8_t i= 0; i < size; ++i) {
+        _port->write(request[i]);
+    }
+
+    _saStat.pktsent++;
+    //AP_SmartAudio::_print_bytes_to_hex_string(request, size);
+}
+
+/**
+ * Reads the response from vtx in the wire
+ * - response_buffer, response buffer to fill in
+ * - inline_buffer_length , used to passthrought the response lenght in case the response where splitted
+ **/
+void AP_SmartAudio::read_response(uint8_t *response_buffer, uint8_t inline_buffer_length)
+{
+    int16_t incoming_bytes_count = _port->available();
```

Comment:
Are you expecting a response here? If so you should keep reading until you get something or timeout - or make the function re-entrant and do this from the thread loop. At the moment you simply give up if there are not enough bytes.

Reply:
Yes this method makes the response reading. Here the splitted response must be captured and completed. I'll check the impl to acomplish well the task.

Reply:
Fixed in db8526f24036d58d4f88554c9750ae65d05869a1